lib/charms/ceph_csi/v0/ceph_client.py

```diff
@@ -0,0 +1,225 @@
+        :type compression_min_blob_size_ssd: Optional[int]
+        :param compression_max_blob_size: Chunks larger than this are broken into
+                                          N * compression_max_blob_size chunks before being
+                                          compressed (unit: bytes).
+        :type compression_max_blob_size: Optional[int]
+        :param compression_max_blob_size_hdd: Chunks larger than this are broken into
+                                              N * compression_max_blob_size_hdd chunks before
+                                              being compressed when destined for rotational media
+                                              (unit: bytes)
+        :type compression_max_blob_size_hdd: Optional[int]
+        :param compression_max_blob_size_ssd: Chunks larger than this are broken into
+                                              N * compression_max_blob_size_ssd chunks before
+                                              being compressed when destined for flash media
+                                              (unit: bytes).
+        :type compression_max_blob_size_ssd: Optional[int]
+        :param group: Group to add pool to
+        :type group: Optional[str]
+        :param max_bytes: Maximum bytes quota to apply
+        :type max_bytes: Optional[int]
+        :param max_objects: Maximum objects quota to apply
+        :type max_objects: Optional[int]
+        :param group_namespace: Group namespace
+        :type group_namespace: Optional[str]
+        :param rbd_mirroring_mode: Pool mirroring mode used when Ceph RBD mirroring is enabled.
+        :type rbd_mirroring_mode: Optional[str]
+        :param weight: The percentage of data that is expected to be contained in the pool from
+                       the total available space on the OSDs. Used to calculate number of
+                       Placement Groups to create for pool.
+        :type weight: Optional[float]
+
+        """
+        self.app_name: Optional[str] = None
+        self.group: Optional[str] = None
+        self.max_bytes: Optional[int] = None
+        self.max_objects: Optional[int] = None
+        self.group_namespace: Optional[str] = None
+        self.rbd_mirroring_mode: str = "pool"
+        self.weight: Optional[float] = None
+        self.compression_algorithm: Optional[str] = None
+        self.compression_mode: Optional[str] = None
+        self.compression_required_ratio: Optional[float] = None
+        self.compression_min_blob_size: Optional[int] = None
+        self.compression_min_blob_size_hdd: Optional[int] = None
+        self.compression_min_blob_size_ssd: Optional[int] = None
+        self.compression_max_blob_size: Optional[int] = None
+        self.compression_max_blob_size_hdd: Optional[int] = None
+        self.compression_max_blob_size_ssd: Optional[int] = None
+
+        for key, value in kwargs.items():
+            _ = self.__getattribute__(key)  # Ensure that key is valid attribute
```

Comment:
I was thinking a lot how to best implement class with 16 attributes which is bit against pep8 (imo in this case it's justified) and I think that `**kwargs` is best solution.

:re defining all arguments: This class is not expected to be instantiated directly, so to retain benefits of explicitly stating all arguments as `__init__` parameters, all it's child classes would have to do the same which would get even more awkward with their own arguments on top.

:re using `self.app_name = kwargs.get("app_name")`: This would not raise an error if `kwargs` would contain unexpected arguments (most likely a typo)

:re using hasattr: I like this, it's more explicit than my current approach.

Reply:
I checked [doc](https://docs.python.org/3.5/library/functions.html#hasattr) and hasattr works by checking if `getattr` raises `AttributeError`. So I think that it might be best to leave this one as is. Because using `hasattr` (which calss `getattr`) only to raise `AttributeError` manually seems like really convoluted way to achieve what `getattr` does anyway.

Reply:
> I was thinking a lot how to best implement class with 16 attributes which is bit against pep8 (imo in this case it's justified) and I think that `**kwargs` is best solution.
> 
> :re defining all arguments: This class is not expected to be instantiated directly, so to retain benefits of explicitly stating all arguments as `__init__` parameters, all it's child classes would have to do the same which would get even more awkward with their own arguments on top.

The child classes could use kwargs to call `super().__init__(**kwargs)`, but this parent class should have all arguments defined. This will also ensure that all arguments have right name.

> 
> :re using `self.app_name = kwargs.get("app_name")`: This would not raise an error if `kwargs` would contain unexpected arguments (most likely a typo)
> 
> :re using hasattr: I like this, it's more explicit than my current approach.

Reply:
I was thinking about using something like this:
```python
if not hasattr(self, key):
  raise ValueError("...")
```